### PR TITLE
dm: add foreign key test for full phase

### LINKED ISSUE #7598: DM support sync table with foreign key
### Is your feature request related to a problem?

Related TiDB Issue: [Support Foreign Key Dev Task#36982](https://github.com/pingcap/tidb/issues/36982)

### Describe the feature you'd like

N/A

### Describe alternatives you've considered

N/A

### Teachability, Documentation, Adoption, Migration Strategy

N/A

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Issue Number: close #7598

#### Tests 

 - Integration test
 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: dm/tests/others_integration_2.txt [hunk 1]
```diff
@@ -9,3 +9,4 @@ http_proxies
 openapi
 duplicate_event
 binlog_parse
+foreign_key
```
Comment (lance6716) [inline dm/tests/others_integration_2.txt:12]: you can add this test at the beginning of this file, so CI will fail more earlier

**Comment by lance6716** [comment]
Seems now we have two problems:

- `set @@global.tidb_enable_foreign_key=1;` should be executed in downstream first, we can do it using `run_sql_tidb "set @@global.tidb_enable_foreign_key=1;"` at the beginning of run() .
- lightning does not respect table FK dependency when creating table.

**Comment by lance6716** [comment]
I make a simple test

```
MySQL [(none)]> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                               |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v6.5.0-alpha
Edition: Community
Git Commit Hash: e205f931aaf0f8b56bdb6a809be3435ef1de764d
Git Branch: heads/refs/tags/v6.5.0-alpha
UTC Build Time: 2022-11-24 14:33:44
GoVersion: go1.19.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

MySQL [(none)]> drop database foreign_key;
Query OK, 0 rows affected (0.177 sec)

MySQL [(none)]> create database foreign_key;
Query OK, 0 rows affected (0.073 sec)

MySQL [(none)]> CREATE TABLE IF NOT EXISTS `foreign_key`.`t1` (`id` INT(11) NOT NULL,PRIMARY KEY(`id`)) ENGINE = InnoDB DEFAULT CHARACTER SET = UTF8MB4 DEFAULT COLLATE = UTF8MB4_BIN;
Query OK, 0 rows affected (0.068 sec)

MySQL [(none)]> CREATE TABLE IF NOT EXISTS `foreign_key`.`t2` (`id` INT(11) NOT NULL,PRIMARY KEY(`id`),CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`id`) REFERENCES `t1`(`id`) ON DELETE CASCADE ON UPDATE CASCADE) ENGINE = InnoDB DEFAULT CHARACTER SET = UTF8MB4 DEFAULT COLLATE = UTF8MB4_BIN;
Query OK, 0 rows affected (0.099 sec)

MySQL [(none)]> use foreign_key;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MySQL [foreign_key]> show create table t2;
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                           |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t2    | CREATE TABLE `t2` (
  `id` int(11) NOT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`id`) REFERENCES `foreign_key`.`t1` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.001 sec)

MySQL [foreign_key]> insert into t1 values(1);
Query OK, 1 row affected (0.002 sec)

MySQL [foreign_key]> insert into t2 values(1);
Query OK, 1 row affected (0.002 sec)

MySQL [foreign_key]> select * from t2;
+----+
| id |
+----+
|  1 |
+----+
1 row in set (0.002 sec)

MySQL [foreign_key]> delete from t1 where id = 1;
Query OK, 1 row affected (0.002 sec)

MySQL [foreign_key]> select * from t2;
+----+
| id |
+----+
|  1 |
+----+
1 row in set (0.002 sec)
```

why ON DELETE CASCADE does not work?

**Comment by crazycs520** [comment]
> e205f931aaf0f8b56bdb6a809be3435ef1de764d

Need set following 2 variables to enable foreign key:

```sql
set @@global.tidb_enable_foreign_key=1;
set @@foreign_key_checks=1;
```

**Review by lance6716 — APPROVED** [review]

**Comment by lance6716** [comment]
/cc @GMHDBJD @buchuitoudegou 

we will fix the syncer phase later by checking `OPTION_NO_FOREIGN_KEY_CHECKS` in `Q_FLAGS2_CODE`, now CI of the full phase is passed

**Review by buchuitoudegou — APPROVED** [review]
lgtm

but if tidb supports foreign key now, should we remove the foreign key check when doing precheck @lance6716

**Comment by lance6716** [comment]
> lgtm
> 
> but if tidb supports foreign key now, should we remove the foreign key check when doing precheck @lance6716

after we fix syncer phase FK support

/merge

**Pull request merged**